File: ENGR2300_AppliedLinearAlgebra/Chapter3-4/PYTHON/Homework/ch3_3_and_4_Homework_Script.py
# Crystal Hollis
# ENGR 2300 Applied Linear Algebra
# Homework 6 Chapters 3.3 and 3.4

# Import libraries
from sympy import symbols, Matrix, solve, Eq, det

# Problem 3: Find the value of k such that A is singular.
#     [0, k, 1],
#     [k, 1, k],
#     [1, k, 0]

# Step 1: Define symbol
k = symbols('k')

# Step 2: Define matrix A with symbolic k
A3 = Matrix([
    [0, k, 1],
    [k, 1, k],
    [1, k, 0]
])

# Step 3: Compute determinant
det_A3 = A3.det()

# Step 4: Solve equation det(A) = 0
solutions = solve(Eq(det_A3, 0), k)
print(solutions) # output: [-sqrt(2)/2, sqrt(2)/2]
# Answer: k = -sqrt(2)/2, sqrt(2)/2

# Problem 5: Find |A|, |B|, AB, and |AB|. Then verify that |A||B| = |AB|.
# A =
#     [-1,  2, 1],
#     [ 1,  0, 1],
#     [ 0,  1, 0]
#
# B =
#     [-1,  0,  0],
#     [ 0,  2,  0],
#     [ 0,  0,  5]

# Define Matrices A and B:
A = Matrix([
    [-1,  2, 1],
    [ 1,  0, 1],
    [ 0,  1, 0]
])

B = Matrix([
    [-1,  0,  0],
    [ 0,  2,  0],
    [ 0,  0,  5]
])

# Compute determinants of A and B:
det_A = A.det()
det_B = B.det()
print("det(A) =", det_A)
print("det(B) =", det_B)

# Multiply AB and compute its determinant:
AB = A * B
print("AB =")
print(AB)
det_AB = AB.det()
print("det(AB) =", det_AB)

# Verify
print("det(A)*det(B) =", det_A * det_B)
print("Are they equal?", det_A * det_B == det_AB)

# Problem 6: Use a determinant to decide whether the matrix is singular or nonsingular.
# A =
#     [0.9,  0.3, -0.5,  0.1],
#     [-1.2, 0.3,  0.6,  0  ],
#     [0.6, -0.7,  0.1,  0  ],
#     [0.6, -0.3,  0.2,  0  ]

# import libraries
import numpy as np

# define Matrix A using numpy
A6 = np.array([
    [0.9,  0.3, -0.5,  0.1],
    [-1.2, 0.3,  0.6,  0  ],
    [0.6, -0.7,  0.1,  0  ],
    [0.6, -0.3,  0.2,  0  ]
])

# Compute Determinant of A using numpy
det_A6 = np.linalg.det(A6)
print("det(A6) =", det_A6)


# Interpret the results: If det_A is exactly zero (or extremely close to zero) the matrix is singular.
# Otherwise, the matrix is nonsingular.
tolerance = 1e-12
if abs(det_A6) < tolerance:
    print("Matrix A6 is SINGULAR")
else:
    print("Matrix A6 is NONSINGULAR")

# Problem 11: Use Cramer's Rule to solve (if possible) the system of linear equations. (If not possible, enter IMPOSSIBLE.)
# A11 =
#     [4,  -2,  3],
#     [2,   2,  5],
#     [8,  -5, -2]
# Constant Terms = [-5, 5, 8]

# Library already imported:
# import sympy
# from sympy import Matrix

# Define the coefficient Matrix A and the constant vector b:
A11 = Matrix([
    [4,  -2,  3],
    [2,   2,  5],
    [8,  -5, -2]
])

b11 = Matrix([-5, 5, 8])

# Compute the determinant of A:
detA11 = A11.det()
print("det(A11) =", detA11)

# Form the matrics A_x, A_y, A_z by replacing one column at a time with b:
# A_x: replace the 1st column of A with b
A11_x = A11.copy()
A11_x[:,0] = b11
detA11_x = A11_x.det()

# A_y: replace the 2nd column of A with b
A11_y = A11.copy()
A11_y[:,1] = b11
detA11_y = A11_y.det()

# A_z: replace the 3rd column of A with b
A11_z = A11.copy()
A11_z[:,2] = b11
detA11_z = A11_z.det()

# Check if det(A) = 0. If zero, the system is not uniquely solvable by Cramer's Rule
if detA11 == 0:
    print("IMPOSSIBLE: System is not uniquely solvable (A is singular).")
else:
    x = detA11_x / detA11
    y = detA11_y / detA11
    z = detA11_z / detA11
    print(f"x = {x}, y = {y}, z = {z}")

# Interpret the result:
# If det(A) ≠ 0, you get unique (x, y, z).
# If det(A) = 0, you must answer “IMPOSSIBLE”
# (or the system might have infinitely many or no solutions, but Cramer’s Rule can’t solve it).

File: ENGR2300_AppliedLinearAlgebra/Chapter3-4/MATLAB/Homework/ch3_3_and_4_Homework_script.m
% Crystal Hollis
% ENGR 2300 Applied Linear Algebra
% A6 Determinants Chapters 3.3 and 3.4

% Problem 7: Find the volume of the tetrahedron having the given vertices:
% p = [5; -5; 1]; q = [3; -2; 4]; r = [1; 1; 1]; s = [0; 0; 1];

% Enter the points as column vectors 
p = [5; -5; 1];
q = [3; -2; 4];
r = [1; 1; 1];
s = [0; 0; 1];

% Form the difference vectors from one vertex, P. 
% This gives three vectors that define the edges of the 
% tetrahedron from P.
PQ = q - p;
PR = r - p;
PS = s - p;

% Construct the 3×3 matrix whose columns are these difference vectors:
% M = [Q - P R - P S - P]. 
M7 = [PQ, PR, PS];

% Compute the determinant of that matrix:
d = det(M7);

% Compute the volume, which is 1/6 of the absolute value of that
% determinant
volume = abs(d) / 6
disp('cubic units')

%%

% Problem 8: Determine whether the points are coplanar.
% P1 = [1; -7;  9]; P2 = [-1; -7;  9]; P3 = [1; -7; -9]; P4 = [-1; -7; -9];

% Enter the points as column vectors:
P1 = [1; -7;  9];
P2 = [-1; -7;  9];
P3 = [1; -7; -9];
P4 = [-1; -7; -9];

% Form the difference vectors from P1:
v1 = P2 - P1;   % vector P1->P2
v2 = P3 - P1;   % vector P1->P3
v3 = P4 - P1;   % vector P1->P4

% Construct a 3x3 matrix using these vectors as columns:
M8 = [v1, v2, v3];

% Compute the determinant:
d = det(M8);
fprintf('Determinant = %.4f\n', d);

% Interpret the result:
% If d == 0 (or extremely close to zero within floating-point tolerance)
% the points are coplanar. Otherwise, not coplanar.
% Since det(M) = 0, the four points are coplanar

%%

% Problem 10: Find the adjoint of the matrix A. Then use the adjoint to 
% find the inverse of A (if possible). 
% (If not possible / Singular, enter IMPOSSIBLE.)
% A = [
%    -6  -3  -5
%     2   4   3
%     0   1  -1]

% Enter matrix A:
A = [
    -6  -3  -5
     2   4   3
     0   1  -1
];


% Compute the determinant:
dA = det(A);
fprintf('det(A) = %f\n', dA);

% Check if det(A) = 0
if abs(dA) < 1e-12
    disp('IMPOSSIBLE (A is singular)');
else
    % If nonzero, we can proceed
    invA = inv(A);
    % Then the adjoint is det(A)*inv(A)
    adjA = dA * invA;
    disp('adj(A) =');
    disp(adjA);
    
    disp('A^-1 =');
    disp(invA);
end

% Interpret the results: if IMPOSSIBLE (A is singular), then
% det(A) = 0 and there is no inverse.
% Otherwise, there are numeric matrices for adj(A) and inverse of A.

%%
% Problem 12: Use the fact that |cA| = cn|A| to evaluate the 
% determinant of the n × n matrix.
% A = [9 12; 3 -6];  this is a 2x2 matrix

A12 = [9 12; 3 -6];
[m, n] = size(A12);
disp([m, n])  % should be [2, 2]

% STEP 1: Factor out the greatest common divisor.
% By hand, the greatest common divisor is 3
% Dimension n = 2
% Reduced matrix  is A / c
c12 = 3;    % greatest common divisor
B12 = A12 / c12;
disp(B12)

% STEP 2: Find the determinant of the matrix found in Step 1.
detB12 = det(B12);
fprintf('det(B12) = %.4f\n', detB12);

% STEP 3: Find the determinant of the original matrix.
n = 2;   % because A is 2x2
detA12_fromFactor = (c12^n) * detB12;
fprintf('det(A12) via factor-out method = %.4f\n', detA12_fromFactor);

% Verify by directly computing det(A) with built-in function
detA12_direct = det(A12);
fprintf('det(A12) direct = %.4f\n', detA12_direct);
